File: 20242Q/poo/laboratorio/tp04/_soluciones/ej3/Rectangle.java
package ar.edu.itba.poo.soltp4.ej3;

public class Rectangle extends Figure {

    protected final Point topLeft, bottomRight;

    public Rectangle(Point topLeft, Point bottomRight) {
        this.topLeft = topLeft;
        this.bottomRight = bottomRight;
    }

    public double base() {
        return Math.abs(topLeft.getX() - bottomRight.getX());
    }

    public double height() {
        return Math.abs(topLeft.getY() - bottomRight.getY());
    }

    @Override
    public double area() {
        return base() * height();
    }

    @Override
    public double perimeter() {
        return (base() + height()) * 2;
    }

    @Override
    public String toString() {
        return "Rectángulo [ %s , %s ]".formatted(topLeft, bottomRight);
    }

}

File: 20242Q/poo/laboratorio/tp09/_soluciones/ej3/rectangle.rb
require_relative 'figure'

class Rectangle < Figure
  attr_reader :top_left_point, :bottom_right_point

  def initialize(top_left_point, bottom_right_point)
    @top_left_point = top_left_point
    @bottom_right_point = bottom_right_point
  end

  def base
    (@top_left_point.x - @bottom_right_point.x).abs
  end

  def height
    (@top_left_point.y - @bottom_right_point.y).abs
  end

  def area
    base * height
  end

  def perimeter
    (base + height) * 2
  end

  def to_s
    "Rectángulo [ #{@top_left_point} , #{@bottom_right_point} ]"
  end

  def ==(other)
    return false unless other.is_a?(Rectangle)
    @top_left_point == other.top_left_point && @bottom_right_point == other.bottom_right_point
  end

  def hash
    [@top_left_point, @bottom_right_point].hash
  end
end
